File: src/main/java/com/fish/extendedae_plus_client/impl/AliasGetter.java
package com.fish.extendedae_plus_client.impl;

import com.fish.extendedae_plus_client.integration.ContextModLoaded;
import com.fish.extendedae_plus_client.util.UtilKeyBuilder;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import dev.emi.emi.api.EmiApi;
import dev.emi.emi.api.recipe.EmiRecipe;
import dev.emi.emi.api.stack.EmiIngredient;
import dev.emi.emi.jemi.JemiRecipe;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.crafting.RecipeHolder;
import net.neoforged.fml.loading.FMLPaths;

import java.io.IOException;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Predicate;

public class AliasGetter {
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
    private static final String CONFIG_RELATIVE = "extendedae_plus/stored_alias.json";
    private static final Map<String, String> ALIASES = new ConcurrentHashMap<>();

    static {
        tryLoadAliases();
    }

    public static synchronized void tryLoadAliases() {
        try {
            var pathFile = FMLPaths.CONFIGDIR.get().resolve(CONFIG_RELATIVE);
            if (!Files.exists(pathFile)) Files.createFile(pathFile);

            var json = Files.readString(pathFile);
            var obj = GSON.fromJson(json, JsonObject.class);
            if (obj == null) {
                ALIASES.clear();
                return;
            }

            var resolvedAlias = new HashMap<String, String>();

            obj.entrySet().forEach(entry -> {
                var typeKey = entry.getKey();
                var aliasValve = entry.getValue();
                if (aliasValve == null || !aliasValve.isJsonPrimitive()) return;

                var aliasName = aliasValve.getAsString();
                if (aliasName == null || aliasName.isBlank()) return;

                resolvedAlias.put(typeKey.toLowerCase(), aliasName);
            });

            ALIASES.clear();
            ALIASES.putAll(resolvedAlias);
        } catch (Throwable ignored) {
        }
    }

    /**
     * 向配置中新增或更新别名映射，并刷新内存映射。
     *
     * @param typeKey 最终搜索关键字（不含冒号），大小写不敏感
     * @param alias  别名
     * @return 是否写入成功
     */
    public static synchronized boolean addOrUpdateAlias(String typeKey, String alias) {
        if (typeKey == null || typeKey.isBlank() || alias == null || alias.isBlank())
            return false;

        try {
            var pathFile = FMLPaths.CONFIGDIR.get().resolve(CONFIG_RELATIVE);
            if (!Files.exists(pathFile)) Files.createFile(pathFile);

            JsonObject obj;
            if (Files.exists(pathFile)) {
                String json = Files.readString(pathFile);
                obj = GSON.fromJson(json, JsonObject.class);
                if (obj == null) obj = new JsonObject();
            } else return false;

            var key = typeKey.trim();

            obj.addProperty(key, alias);
            Files.writeString(pathFile, GSON.toJson(obj));

            ALIASES.put(key.toLowerCase(), alias);
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public static synchronized int removeAliases(String alias) {
        if (alias == null) return 0;

        var target = alias.trim();
        if (target.isBlank()) return 0;

        try {
            var pathFile = FMLPaths.CONFIGDIR.get().resolve(CONFIG_RELATIVE);
            if (!Files.exists(pathFile)) {
                Files.createFile(pathFile);
                return 0;
            }

            var json = Files.readString(pathFile);
            var obj = GSON.fromJson(json, JsonObject.class);
            if (obj == null) return 0;

            var toRemove = new ArrayList<String>();
            obj.entrySet().forEach(entry -> {
                var aliasValue = entry.getValue();
                if (aliasValue == null || !aliasValue.isJsonPrimitive()) return;

                var aliasName = aliasValue.getAsString();
                if (target.equalsIgnoreCase(aliasName))
                    toRemove.add(entry.getKey().toLowerCase());
            });
            if (toRemove.isEmpty()) return 0;

            toRemove.forEach(ALIASES::remove);
            toRemove.forEach(obj::remove);
            Files.writeString(pathFile, GSON.toJson(obj));
            return toRemove.size();
        } catch (IOException e) {
            return 0;
        }
    }

    public static String findMapping(String key) {
        if (key == null || key.isBlank()) return null;

        if (ALIASES.containsKey(key.toLowerCase()))
            return ALIASES.get(key.toLowerCase());

        return null;
    }

    /// 收集到处理配方的关键词（按优先级排序）
    public static volatile List<KeywordGroup> recipeKeywords = new ArrayList<>() {
        @Override
        public void clear() {
            keyUsed = false;
            super.clear();
        }
    };
    public static boolean keyUsed = false;

    public static List<KeywordGroup> getRecipeKeywords() {
        keyUsed = true;
        recipeKeywords.sort(Comparator
                .comparing(KeywordGroup::isMapped).reversed()
                .thenComparing(KeywordGroup::getPriority, Comparator.reverseOrder()));
        return recipeKeywords;
    }

    public static void collectRecipeKeyword(String name, int priority, boolean findMapping) {
        if (keyUsed) recipeKeywords.clear();

        var group = KeywordGroup.literal(name);
        group.setPriority(priority);
        if (findMapping) group.findMapping(true);
        recipeKeywords.add(group);
    }

    /// @param recipe (J)EmiRecipe或RecipeHolder
    public static void tryCollectKeywords(Object recipe) {
        recipeKeywords.clear();
        if (recipe == null) return;
        var keys = new HashMap<String, Integer>();

        if (ContextModLoaded.emi.isLoaded()) {
            List<EmiIngredient> workstations = new ArrayList<>();
            Component categoryName = Component.empty();

            if (recipe instanceof JemiRecipe<?> jemiRecipe) {
                workstations = EmiApi.getRecipeManager().getWorkstations(jemiRecipe.recipeCategory);
                categoryName = jemiRecipe.category.getTitle();

                keys.put(jemiRecipe.category.getTitle().getString(), 3);
                if (jemiRecipe.originalId != null) {
                    keys.put(jemiRecipe.originalId.toString().split("/")[0], 2);
                    keys.put(jemiRecipe.originalId.getPath().split("/")[0], 1);
                }
            } else if (recipe instanceof EmiRecipe emiRecipe) {
                workstations = EmiApi.getRecipeManager().getWorkstations(emiRecipe.getCategory());
                categoryName = emiRecipe.getCategory().getName();

                keys.put(emiRecipe.getCategory().getName().getString(), 3);
                if (emiRecipe.getId() != null) {
                    keys.put(emiRecipe.getId().toString().split("/")[0], 2);
                    keys.put(emiRecipe.getId().getPath().split("/")[0], 1);
                }
            }

            if (!workstations.isEmpty()) {
                var workstationKeys = new ArrayList<String>();
                workstations.reversed().forEach(ingredient -> ingredient.getEmiStacks().reversed()
                        .forEach(stack -> {
                            workstationKeys.add(stack.getName().getString());
                            workstationKeys.add(stack.getName().toString());
                        })
                );

                var groupWorkstation = new KeywordGroup(workstationKeys,
                        UtilKeyBuilder.of(UtilKeyBuilder.keywordGroup)
                                .addStr("workstations")
                                .args(categoryName.getString())
                                .build());
                groupWorkstation.setPriority(4);
                groupWorkstation.findMapping(false);

                recipeKeywords.add(groupWorkstation);
            }
        }

        if (recipe instanceof RecipeHolder<?> recipeHolder){
            keys.put(recipeHolder.id().toString().split("/")[0], 2);
            keys.put(recipeHolder.id().getPath().split("/")[0], 1);
        }

        keys.entrySet().stream()
                .filter(entry -> !entry.getKey().isBlank())
                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                .forEach(entry -> collectRecipeKeyword(entry.getKey(), entry.getValue(), true));
    }

    public static class KeywordGroup {
        public static final KeywordGroup EMPTY = KeywordGroup.literal("");

        private static HashMap<String, KeywordGroup> literalGroups = new HashMap<>();

        private final List<String> keywords = new ArrayList<>();
        private Component description;
        private boolean mapped = false;
        private int priority = 0;

        public KeywordGroup(Collection<String> keywords, Component groupDescription) {
            this.keywords.addAll(keywords);
            this.description = groupDescription;
        }

        public static KeywordGroup literal(String value) {
            if (literalGroups == null) literalGroups = new HashMap<>();
            return literalGroups.computeIfAbsent(value, $ ->
                    new KeywordGroup(List.of(value), Component.empty()));
        }

        public boolean matches(String nameKey, String i18nKey) {
            if (this.keywords.stream().map(String::isBlank).allMatch(Predicate.isEqual(true))) return true;

            boolean descMatched = nameMatches(this.description.getString(), nameKey);
            boolean keywordsMatched = keywords.stream().anyMatch(
                    key -> nameMatches(key, nameKey) || i18nKeyMatches(key, i18nKey));

            return descMatched || keywordsMatched;
        }

        private static boolean nameMatches(String matchKey, String searchKey) {
            if (matchKey == null || matchKey.isBlank()) return false;
            if (searchKey == null || searchKey.isBlank()) return true;

            if (ContextModLoaded.jech.isLoaded()) {
                try {
                    Method jecContains = Class.forName("me.towdium.jecharacters.utils.Match")
                            .getMethod("contains", CharSequence.class, CharSequence.class);
                    Object result = jecContains.invoke(null, matchKey, searchKey);
                    if (result instanceof Boolean resultBool) return resultBool;
                } catch (Throwable ignore) {
                }
            }

            return matchKey.toLowerCase().contains(searchKey.toLowerCase()) ||
                    searchKey.toLowerCase().contains(matchKey.toLowerCase());
        }

        private static boolean i18nKeyMatches(String matchKey, String searchKey) {
            if (matchKey == null || matchKey.isBlank() ||
                    searchKey == null || searchKey.isEmpty()) return false;
            return matchKey.toLowerCase().contains(searchKey.toLowerCase()) ||
                    searchKey.toLowerCase().contains(matchKey.toLowerCase());
        }

        public Component getDescription() {
            if (this.description.getString().isEmpty())
                return Component.literal(this.keywords.getFirst());
            else return this.description;
        }

        public void findMapping(boolean mappingKeywords) {
            var mappedDesc = AliasGetter.findMapping(this.description.getString());
            if (mappedDesc != null && !mappedDesc.isBlank()) {
                this.description = Component.literal(mappedDesc);
                this.mapped = true;
            }

            if (!mappingKeywords) return;

            AtomicBoolean mapped = new AtomicBoolean(false);
            var mappedList = this.keywords.stream().map(keyword -> {
                var mappedKey = AliasGetter.findMapping(keyword);
                if (mappedKey != null) {
                    mapped.set(true);
                    return mappedKey;
                } else return keyword;
            }).toList();
            if (mapped.get()) this.mapped = true;

            this.keywords.clear();
            this.keywords.addAll(mappedList);
        }

        public boolean isMapped() {
            return mapped;
        }

        public int getPriority() {
            return priority;
        }

        public void setPriority(int priority) {
            this.priority = priority;
        }

        public boolean isEmpty() {
            return keywords.isEmpty();
        }

        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof KeywordGroup other)) return false;
            return this.keywords.equals(other.keywords);
        }
    }
}

File: src/main/kotlin/com/fish/extendedae_plus_client/impl/AliasGetter.kt
package com.fish.extendedae_plus_client.impl

import com.electronwill.nightconfig.core.Config
import com.electronwill.nightconfig.core.file.FileConfig
import com.electronwill.nightconfig.core.file.FileConfig.builder
import com.electronwill.nightconfig.toml.TomlFormat.instance
import com.fish.extendedae_plus_client.integration.ContextModLoaded
import com.fish.extendedae_plus_client.util.UtilKeyBuilder
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import dev.emi.emi.api.EmiApi
import dev.emi.emi.api.recipe.EmiRecipe
import dev.emi.emi.api.stack.EmiIngredient
import dev.emi.emi.api.stack.EmiStack
import dev.emi.emi.jemi.JemiRecipe
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.contents.PlainTextContents
import net.minecraft.network.chat.contents.TranslatableContents
import net.minecraft.world.item.crafting.RecipeHolder
import net.neoforged.fml.loading.FMLPaths
import java.nio.file.Files
import java.util.function.Predicate
import kotlin.concurrent.Volatile
import kotlin.io.path.deleteIfExists
import kotlin.io.path.exists

object AliasGetter {
    private val pathConfig = FMLPaths.CONFIGDIR.get().resolve("extendedae_plus/stored_alias.toml")
    private var pathConfigOld = pathConfig.parent.resolve("stored_alias.json")
    private val config: FileConfig = builder(pathConfig, instance()).autosave().autoreload().build()

    init {
        this.config.load()

        if (this.pathConfigOld.exists())
            this.tryConvertConfig()
    }

    fun closeConfig() {
        this.config.close()
    }

    fun tryConvertConfig() {
        val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
        val obj = gson.fromJson(Files.readString(this.pathConfigOld), JsonObject::class.java) ?: return

        obj.entrySet().stream()
            .filter { entry -> entry.value?.isJsonPrimitive ?: false && !entry.value?.asString.isNullOrBlank() }
            .forEach { entry -> this.config.set(entry.key, entry.value.asString) }
        this.config.save()
        this.pathConfigOld.deleteIfExists()
    }

    /**
     * 向配置中新增或更新别名映射，并刷新内存映射。
     * 
     * @param typeKey 最终搜索关键字（不含冒号），大小写不敏感
     * @param alias  别名
     * @return 是否写入成功
     */
    @Synchronized
    fun addOrUpdateAlias(typeKey: String?, alias: String?): Boolean {
        if (typeKey.isNullOrBlank() || alias.isNullOrBlank()) return false

        this.config.set<String>(listOf(typeKey.lowercase()), alias)
        return true
    }

    /** @return 删除别名的数量 */
    @Synchronized
    fun removeAliases(alias: String?): Int {
        if (alias.isNullOrBlank()) return 0

        val target = alias.trim()
        if (target.isBlank()) return 0

        val toRemove = this.config.entrySet().stream()
            .filter { entry -> entry.getValue<Any>().toString().equals(target, true) }
            .map(Config.Entry::getKey)
            .toList()
        toRemove.forEach { value -> this.config.remove(value) }
        return toRemove.size
    }

    fun findMapping(key: String?): String? {
        if (key.isNullOrBlank()) return null

        return this.config.get(listOf(key.lowercase()))
    }

    /** 收集到处理配方的关键词（按优先级排序） */
    @Volatile
    private var recipeKeywords: MutableList<KeywordGroup> = ArrayList<KeywordGroup>()

    fun clearRecipeKeywords() {
        recipeKeywords.clear()
    }

    fun getRecipeKeywords(): MutableList<KeywordGroup> {
        recipeKeywords.sortWith(
            Comparator
                .comparing(KeywordGroup::isMapped).reversed()
                .thenComparing(KeywordGroup::priority, Comparator.reverseOrder())
        )
        return recipeKeywords
    }

    fun collectRecipeKeyword(name: String, priority: Int, findMapping: Boolean) {
        val group = KeywordGroup.literal(name)
        group.priority = priority
        if (findMapping) group.findMapping(true)
        recipeKeywords.add(group)
    }

    /** @param recipe (J)EmiRecipe或RecipeHolder
     */
    @JvmStatic
    fun tryCollectKeywords(recipe: Any?) {
        recipeKeywords.clear()
        if (recipe == null) return
        val keys = HashMap<String, Int>()

        if (ContextModLoaded.emi.isLoaded) {
            var workstations: MutableList<EmiIngredient> = ArrayList<EmiIngredient>()
            var categoryName: Component = Component.empty()

            if (recipe is JemiRecipe<*>) {
                workstations = EmiApi.getRecipeManager().getWorkstations(recipe.recipeCategory)
                categoryName = recipe.category.title

                keys[recipe.category.title.string] = 3
                if (recipe.originalId != null) {
                    keys[recipe.originalId.toString().split("/")[0]] = 2
                    keys[recipe.originalId.path.split("/")[0]] = 1
                }
            } else if (recipe is EmiRecipe) {
                workstations = EmiApi.getRecipeManager().getWorkstations(recipe.category)
                categoryName = recipe.category.name

                keys[recipe.category.name.string] = 3
                if (recipe.id != null) {
                    keys[recipe.id.toString().split("/")[0]] = 2
                    keys[recipe.id!!.path.split("/")[0]] = 1
                }
            }

            if (!workstations.isEmpty()) {
                val workstationKeys = ArrayList<String>()
                workstations.reversed().forEach { ingredient: EmiIngredient ->
                    ingredient.emiStacks.reversed()
                        .forEach { stack: EmiStack ->
                            val name = stack.name
                            workstationKeys.add(name.string)

                            var key: String? = null
                            val contents = name.contents
                            if (contents is PlainTextContents) key = contents.text()
                            else if (contents is TranslatableContents) key = contents.key
                            if (key == null) return@forEach
                            workstationKeys.add(key)
                        }
                }

                val groupWorkstation = KeywordGroup(
                    workstationKeys,
                    UtilKeyBuilder.of(UtilKeyBuilder.keywordGroup)
                        .addStr("workstations")
                        .args(categoryName.string)
                        .build()
                )
                groupWorkstation.priority = 4
                groupWorkstation.findMapping(false)

                recipeKeywords.add(groupWorkstation)
            }
        }

        if (recipe is RecipeHolder<*>) {
            keys[recipe.id().toString().split("/")[0]] = 2
            keys[recipe.id().path.split("/")[0]] = 1
        }

        keys.entries.stream()
            .filter { entry -> entry.key.isBlank() }
            .sorted { a, b -> Comparator.reverseOrder<Int>().compare(a.value, b.value) }
            .forEach { entry ->
                collectRecipeKeyword(
                    entry.key,
                    entry.value,
                    true
                )
            }
    }

    class KeywordGroup(keywords: MutableCollection<String>, groupDescription: Component) {
        private val keywords: MutableList<String> = ArrayList<String>()
        private var description: Component
        var isMapped: Boolean = false
            private set
        var priority: Int = 0

        init {
            this.keywords.addAll(keywords)
            this.description = groupDescription
        }

        fun matches(nameKey: String, i18nKey: String): Boolean {
            if (this.keywords.stream()
                    .map(String::isBlank)
                    .allMatch(Predicate.isEqual(true))
            ) return true

            return nameMatches(this.description.string, nameKey)
                    || this.keywords.stream().anyMatch { key ->
                nameMatches(key, nameKey)
                        || i18nKeyMatches(key, i18nKey)
            }
        }

        fun getDescription(): Component {
            return if (this.description.string.isEmpty())
                Component.literal(this.keywords[0])
            else this.description
        }

        fun findMapping(mappingKeywords: Boolean) {
            val mappedDesc = findMapping(this.description.string)
            if (!mappedDesc.isNullOrBlank()) {
                this.description = Component.literal(mappedDesc)
                this.isMapped = true
            }

            if (!mappingKeywords) return

            val mapped = booleanArrayOf(false)
            val mappedList = this.keywords.stream().map { keyword ->
                val mappedKey = findMapping(keyword)
                if (mappedKey != null) {
                    mapped[0] = true
                    return@map mappedKey
                } else return@map keyword
            }.toList()
            if (mapped[0]) this.isMapped = true

            this.keywords.clear()
            this.keywords.addAll(mappedList)
        }

        val isEmpty: Boolean
            get() = keywords.isEmpty()

        override fun equals(other: Any?): Boolean {
            if (other !is KeywordGroup) return false
            return this.keywords == other.keywords
        }

        companion object {
            private var literalGroups: HashMap<String, KeywordGroup>? = null

            fun literal(value: String): KeywordGroup {
                if (literalGroups == null) literalGroups = HashMap<String, KeywordGroup>()
                return literalGroups!!.computeIfAbsent(value) { _ ->
                    KeywordGroup(
                        mutableListOf(value),
                        Component.empty()
                    )
                }
            }

            private fun nameMatches(matchKey: String?, searchKey: String?): Boolean {
                if (matchKey.isNullOrBlank()) return false
                if (searchKey.isNullOrBlank()) return true

                var jechMatches = false
                if (ContextModLoaded.jech.isLoaded) {
                    try {
                        val methodContains = Class.forName("me.towdium.jecharacters.utils.Match")
                            .getMethod("contains", String::class.java, CharSequence::class.java)
                        jechMatches = methodContains.invoke(
                            null, searchKey.lowercase(), matchKey.lowercase()
                        ) as Boolean
                    } catch (_: Throwable) {
                    }
                }

                return jechMatches
                        || matchKey.lowercase().contains(searchKey.lowercase())
                        || searchKey.lowercase().contains(matchKey.lowercase())
            }

            private fun i18nKeyMatches(matchKey: String?, searchKey: String?): Boolean {
                if (matchKey.isNullOrBlank() || searchKey.isNullOrEmpty()) return false
                return matchKey.lowercase().contains(searchKey.lowercase()) ||
                        searchKey.lowercase().contains(matchKey.lowercase())
            }
        }
    }
}
